Add QPAD simulation package to conda-forge

tdalichaouch opened the pull request:
<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with a PR, please let the
right people know. There are language-specific teams for reviewing recipes.

| Language        | Name of review team           |
| --------------- | ----------------------------- |
| python          | `@conda-forge/help-python`    |
| python/c hybrid | `@conda-forge/help-python-c`  |
| r               | `@conda-forge/help-r`         |
| java            | `@conda-forge/help-java`      |
| nodejs          | `@conda-forge/help-nodejs`    |
| c/c++           | `@conda-forge/help-c-cpp`     |
| perl            | `@conda-forge/help-perl`      |
| Julia           | `@conda-forge/help-julia`     |
| ruby            | `@conda-forge/help-ruby`      |
| Rust            | `@conda-forge/help-rust`      |
| Go              | `@conda-forge/help-go`        |
| other           | `@conda-forge/staged-recipes` |

Once the PR is ready for review, please mention one of the teams above in a
new comment. i.e. `@conda-forge/help-some-language, ready for review!`
Then, a bot will label the PR as 'review-requested'.

Due to GitHub limitations, first time contributors to conda-forge are unable
to ping conda-forge teams directly, but you can [ask a bot to ping the team][1]
using a special command in a comment on the PR to get the attention of the
`staged-recipes` team. You can also consider asking on our [Zulip chat][2]
if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://conda-forge.zulipchat.com

All apologies in advance if your recipe PR does not receive prompt attention.
This is a high volume repository and the reviewers are volunteers. Review times
vary depending on the number of reviewers on a given language team and may be days
or weeks. We are always looking for more staged-recipe reviewers. If you are
interested in volunteering, please contact a member of @conda-forge/core.
We'd love to have the help!
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/qpad/meta.yaml```) and found some lint.

Here's what I've got...


For **recipes/qpad/meta.yaml**:

 * ❌ The top level meta keys are in an unexpected order. Expecting ['package', 'source', 'build', 'requirements', 'test', 'about', 'extra'].
 * ❌ The home item is expected in the about section.
 * ❌ Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form. See lines [17]
 * ❌ This recipe is using a compiler, which now requires adding a build dependence on `{{ stdlib("c") }}` as well. Note that this rule applies to each output of the recipe using a compiler. For further details, please see https://github.com/conda-forge/conda-forge.github.io/issues/2102.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/16275206340](https://github.com/conda-forge/conda-forge-webservices/actions/runs/16275206340). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/qpad/meta.yaml```) and found some lint.

Here's what I've got...


For **recipes/qpad/meta.yaml**:

 * ❌ Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form. See lines [46]


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/16276063352](https://github.com/conda-forge/conda-forge-webservices/actions/runs/16276063352). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/qpad/meta.yaml```) and found it was in an excellent condition.

tdalichaouch commented:
@conda-forge/stages-recipes, ready for review!

tdalichaouch commented:
@conda-forge/staged_recipes, ready for review!

danielnachun commented on recipes/qpad/meta.yaml:
```diff
@@ -0,0 +1,62 @@
+
+
+source:
+  url: https://github.com/UCLA-Plasma-Simulation-Group/QPAD/archive/heads/conda-forge.tar.gz # https://github.com/UCLA-Plasma-Simulation-Group/QPAD
```
Unfortunately we can't accept packages that don't use tagged releases with standard version numbers.  Please ask the upstream to tag a new release from the main branch, or see if it can still build from the latest tagged release fro 2021..  Also please be aware that version updates the feedstock must also come from tagged releases in the future, so the upstream will need to continue tagging new releases for us to be able to update the package.

tdalichaouch replied:
We created a new tagged release v1.6.0 (https://github.com/UCLA-Plasma-Simulation-Group/QPAD/releases/tag/v1.6.0)  and updated the yaml file to build from it. 

All the checks have passed with this new update.

danielnachun replied:
Thank you for addressing this promptly!  My concerns about versioning are now satisfied.

danielnachun commented on recipes/qpad/meta.yaml:
```diff
@@ -0,0 +1,62 @@
+
+test:
+  commands:
+    - test -f $PREFIX/bin/qpad.e  # check that qpad.e exists
```
Did you use this test because running `qpad.e` produces an error?  If there isn't a command like `qpad.e --help` that will return 0, you can instead run a command that will return an error and use something like this to capture the error output:
```
qpad.e 2>&1 | grep -q "some line from error output"
```
That will redirect the error output to standard output, and then check if that error output is what is supposed to be.  This is preferable to just testing for the existence of the file because if there is a segmentation fault or a missing DSO, a different error would be produced and the test would fail.

tdalichaouch replied:
We've added a test with the executable qpad.e as recommended and included a conda build for version pinning. All the tests have passed.

danielnachun reviewed (changes requested):
Can you clarify if this package can only work with OpenMPI, or does it work with other MPI implementations?

tdalichaouch commented on recipes/qpad/meta.yaml:
```diff
@@ -0,0 +1,72 @@
+    - cp input_file/blowout_regime/qpinput_tri-gaussian.json qpinput.json
+    - test -f $PREFIX/bin/qpad.e  # check that qpad.e exists
+    - test -x $PREFIX/bin/qpad.e  # check that qpad.e is executable
+    - test -f qpinput.json  # check that input file exists
```
Removed lines in latest commit

tdalichaouch replied:
Added mpich to conda config file

tdalichaouch commented on recipes/qpad/meta.yaml:
```diff
@@ -0,0 +1,72 @@
+about:
+  license: GPL-3.0-or-later
+  license_file: LICENSE
+  license_family: GPL
```
Removed line

danielnachun commented on recipes/qpad/conda_build_config.yaml:
```diff
@@ -0,0 +1,11 @@
+mpi:
+  - openmpi
+  - mpich
+mpich:
+  - 4.3.1
+openmpi: 
+  - 5.0.8 
+hdf5:
+  - 1.14.6
```
```suggestion
```
Let's delete these explicit versions for now.  These are the defaults we use anyway in conda-forge, and it's better to let our bots handle version updates for dependencies than manually setting them here.  If and when some version constraining is necessary we can do it in the feedstock.

tdalichaouch replied:
I removed explicit version for everything except hypre >=2.32.0. Previous versions (below 2.32.0) are built without necessary c-to-fortran interfacing needed for the QPAD package. 

While Hypre >=2.32.0 is compatible with the current global pin for mpich 4, it migrated to openmpi 5 while the [global pin is still openmpi 4 ](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/main/recipe/conda_build_config.yaml). Thus, it will fail the CI check without explicit version pinning to openmpi 5. To avoid this issue, I have temporarily disabled the openmpi build.

The repo is ready for re-review.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/qpad/meta.yaml```) and found it was in an excellent condition.

I do have some suggestions for making it better though...


For **recipes/qpad/meta.yaml**:

 * ℹ️ top-level output has some malformed specs:
- In section host: `hypre >=2.32.0 * mpi_mpi*`
- In section run: `hypre >=2.32.0 * mpi_mpi*`
Requirement spec fields should match the syntax `name [version [build]]`to avoid known issues in conda-build. For example, instead of `name =version=build`, use `name version.* build`. There should be no spaces between version operators and versions either: `python >= 3.8` should be `python >=3.8`.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/17001964659](https://github.com/conda-forge/conda-forge-webservices/actions/runs/17001964659). Examine the logs at this URL for more detail.</sub>

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/qpad/meta.yaml```) and found it was in an excellent condition.

tdalichaouch commented:
> Only two small changes needed now and then we should be set.

Removed the MPI run dependency. 

I tried removing libgomp but got an error that the library was missing so I added it back in.

It is ready for re-review.

danielnachun commented:
Could you re-remove `libgomp` from the build requirements and add `V=1` to the `make`?  I want to see what is going on here because the linker and preprocesser shouldn't be using `libgomp` from `$BUILD_PREFIX`.  Unfortunately Azure pipelines deletes the output when you push a new commit so I can't see the previous failed run.

tdalichaouch commented:
> Could you re-remove `libgomp` from the build requirements and add `V=1` to the `make`? I want to see what is going on here because the linker and preprocesser shouldn't be using `libgomp` from `$BUILD_PREFIX`. Unfortunately Azure pipelines deletes the output when you push a new commit so I can't see the previous failed run.

I pushed a new commit with verbose build & `libgomp` removed.

danielnachun commented:
Thanks this confirmed what suspected.  Please put `libgomp` in the `host:` section as a Linux-only dependency and see if things work.

tdalichaouch commented:
> Thanks this confirmed what suspected. Please put `libgomp` in the `host:` section as a Linux-only dependency and see if things work.

I moved `libgomp` to the `host` section. All the checks seem to pass except for `staged-recipes linter` which raises a GithubException Error. This error is being thrown even on previous build which have passed all checks. Have you seen this before?

carterbox commented on recipes/qpad/meta.yaml:
```diff
@@ -0,0 +1,67 @@
+    - make
+    - {{ mpi }}
+    - llvm-openmp  # [osx]
+    - libgomp      # [linux]
```
@danielnachun Our docs direct users of OpenMP to explicitly depend on it.

https://conda-forge.org/docs/maintainer/knowledge_base/#openmp

traversaro commented:
> All checks have passed and the repo is ready to be merged. Please let me know if there is anything else. @danielnachun
> 
> > Thanks this confirmed what suspected. Please put `libgomp` in the `host:` section as a Linux-only dependency and see if things work.

Sorry for the late review @tdalichaouch, it seems that the CI is now red, can you check? Thanks a lot!

tdalichaouch commented:
> > All checks have passed and the repo is ready to be merged. Please let me know if there is anything else. @danielnachun
> > > Thanks this confirmed what suspected. Please put `libgomp` in the `host:` section as a Linux-only dependency and see if things work.
> 
> Sorry for the late review @tdalichaouch, it seems that the CI is now red, can you check? Thanks a lot!

Hi @traversaro, I removed the chmod line which was causing the CI to be red (the build should automatically set the necessary permissions anyway). I also updated the recipe to use the newest release v1.6.2 and compile two binaries with and without openPMD support. 

Everything appears to check out now and the CIs are green. It is now ready for a final review.

danielnachun reviewed (changes requested):
I missed an issue on macOS earlier - once that is made we should be ready to merge at last.
